models/gif_psc_exp.cpp

```diff
@@ -0,0 +1,469 @@
+
+
+  if ( tau_ex_ <= 0 || tau_in_ <= 0 )
+    throw BadProperty( "Synapse time constants must be strictly positive." );
+}
+
+void
+nest::gif_psc_exp::State_::get( DictionaryDatum& d, const Parameters_& p ) const
+{
+  def< double >( d, names::V_m, y3_ );  // Membrane potential
+  def< double >( d, names::E_sfa, q_ ); // Adaptive threshold potential
+}
+
+void
+nest::gif_psc_exp::State_::set( const DictionaryDatum& d, const Parameters_& p )
+{
+  updateValue< double >( d, names::V_m, y3_ );
+  updateValue< double >( d, names::E_sfa, q_ );
+  initialized_ = false; // vectors of the state should be initialized with new parameter set.
+}
+
+nest::gif_psc_exp::Buffers_::Buffers_( gif_psc_exp& n )
+  : logger_( n )
+{
+}
+
+nest::gif_psc_exp::Buffers_::Buffers_( const Buffers_&, gif_psc_exp& n )
+  : logger_( n )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::gif_psc_exp::gif_psc_exp()
+  : Archiving_Node()
+  , P_()
+  , S_()
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+nest::gif_psc_exp::gif_psc_exp( const gif_psc_exp& n )
+  : Archiving_Node( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::gif_psc_exp::init_state_( const Node& proto )
+{
+  const gif_psc_exp& pr = downcast< gif_psc_exp >( proto );
+  S_ = pr.S_;
+  // S_.r_ref_ = Time(Time::ms(P_.t_ref_remaining_)).get_steps();
+}
+
+void
+nest::gif_psc_exp::init_buffers_()
+{
+  B_.spikes_ex_.clear(); // includes resize
+  B_.spikes_in_.clear(); // includes resize
+  B_.currents_.clear();  //!< includes resize
+  B_.logger_.reset();    //!< includes resize
+  Archiving_Node::clear_history();
+}
+
+void
+nest::gif_psc_exp::calibrate()
+{
+
+  B_.logger_.init();
+
+  V_.h_ = Time::get_resolution().get_ms();
+  V_.rng_ = kernel().rng_manager.get_rng( get_thread() );
+
+  V_.P11ex_ = std::exp( -V_.h_ / P_.tau_ex_ );
+  V_.P11in_ = std::exp( -V_.h_ / P_.tau_in_ );
+
+  double_t tau_m = P_.c_m_ / P_.g_L_;
+
+  // these are determined according to a numeric stability criterion
+  V_.P21ex_ = propagator_32( P_.tau_ex_, tau_m, P_.c_m_, V_.h_ );
+  V_.P21in_ = propagator_32( P_.tau_in_, tau_m, P_.c_m_, V_.h_ );
+
+
+  V_.P33_ = std::exp( -V_.h_ / tau_m );
+  V_.P30_ = 1 / P_.c_m_ * ( 1 - V_.P33_ ) * tau_m;
+  V_.P31_ = ( 1 - V_.P33_ );
+
+  V_.RefractoryCounts_ = Time( Time::ms( P_.t_ref_ ) ).get_steps();
+  assert( V_.RefractoryCounts_ >= 0 ); // since t_ref_ >= 0, this can only fail in error
+
+  // initializing internal state
+  if ( !S_.initialized_ )
+  {
+    for ( uint_t i = 0; i < P_.tau_sfa_.size(); i++ )
+    {
+      V_.Q33_.push_back( std::exp( -V_.h_ / P_.tau_sfa_[ i ] ) );
+      S_.q_sfa_elems_.push_back( 0.0 );
```

Comment:
The `sfa_elems_` depends on `Q33_` (`P_sfa_` in the new version). They should have same size and each time `P_sfa_` gets new values, we should reset the values of `sfa_elems_`. A similar relation exists between `stc_elems_` and `P_stc_`. Therefore, I decided to reconstruct the sfa_elems_ and stc_elems_, after calibration of `P_sfa_` and `P_stc_`.